packages/api/src/car.js

```diff
@@ -129,27 +121,36 @@ export async function carGet (request, env, ctx) {
  * @param {import('./index').Ctx} ctx
  */
 export async function carPost (request, env, ctx) {
+  let blob = await request.blob()
+  // Ensure car blob.type is set; it is used by the cluster client to set the foramt=car flag on the /add call.
+  blob = blob.slice(0, blob.size, 'application/car')
+
+  return handleCarUpload(request, env, ctx, blob)
+}
+
+/**
+ * Post a CAR file.
+ *
+ * @param {import('./user').AuthenticatedRequest} request
+ * @param {import('./env').Env} env
+ * @param {import('./index').Ctx} ctx
+ * @param {Blob} car
+ */
+export async function handleCarUpload (request, env, ctx, car, uploadType = 'Car') {
   const { user, authToken } = request.auth
   const { headers } = request
 
-  let name = headers.get('x-name')
-  if (!name || typeof name !== 'string') {
-    name = `Upload at ${new Date().toISOString()}`
-  }
+  // throws if CAR is invalid by our standards
+  const { size: dagSize } = await carStat(car)
 
-  const blob = await request.blob()
-  const bytes = new Uint8Array(await blob.arrayBuffer())
-  const stat = await carStat(bytes)
-
-  // Ensure car blob.type is set; it is used by the cluster client to set the foramt=car flag on the /add call.
-  const content = blob.slice(0, blob.size, 'application/car')
-
-  const { cid } = await env.cluster.add(content, {
-    metadata: { size: content.size.toString() },
+  // We can't use the `bytes` from cluster where it's a unixfs root partial, as it'll be a shard of the total dag size.
+  // We can't use the `size` from cluster as it's the unixfs FileSize which is 0 for directories, and is not set for raw encoded files, only dag-pb ones.
```

Comment:
Maybe we can add this knowledge to the function where we compute the dag size. What do you think?
Reading through the code, it seems strange this placement as it is not clear what it relates to